### PR TITLE
[profiler] When record_inputs=True, record scalar lists of length <= 30

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #100593

Many ops take as inputs scalars or scalar lists which are important to understand the properties of the op. For example, convolution ops' behavior and output shapes often depend on padding and strides, which are provided as scalars of lists of scalars. This will record scalar lists when record_inputs=True.

Details:
During collection (and this was true before this PR as well), we serialize values and tensor metadata into an InputOutputEncoder. After collection occurs, we deserialize these values to attach the information to each of the events.

This PR does this:
- Adds support for serializing scalar lists during collection / serialization
- Adds an extra field called "Concrete Args"
- Splits up the deserialization process into two steps - one for generating "input shapes" and one for generating "concrete args". We split up input shapes and concrete args to avoid interrupting any previous workflows that relied on the specific data in the input shapes category; additionally, it's just a better description. Note that single scalars will remain in the "input shapes" category as they were already in that category in the past.

Differential Revision: [D45798431](https://our.internmc.facebook.com/intern/diff/D45798431)

### CODE REVIEW

File: torch/csrc/profiler/collection.h [hunk 1]
```diff
@@ -146,6 +148,7 @@ struct ExtraFields<EventType::TorchOp> : TorchOpBasicFields {
   uint64_t correlation_id_;
   time_t end_time_ns_;
   std::vector<op_input_t> inputs_;
+  std::vector<op_input_t> concrete_inputs_;
```
Comment (davidberard98) [inline torch/csrc/profiler/collection.h:151]: Based on some earlier discussion 1-2 months ago, we concluded that we should separate "input dims" and "concrete inputs" in case there are things that rely on "input dims".

Does it need to be separated at this level, or could it be separated just at the level of the kineto traces? (e.g. the changes in profiler_kineto.cpp)? i.e., would it be okay to put all this info into `inputs_` or is it good to keep it separated like I did in this PR?

**Review by aaronenyeshi — COMMENTED** [review]
Overall the patch looks great, I am mostly concerned about the overhead to collection since this will add more serialization steps. At the moment, most profiling sessions will enable record_shapes=True.

Would it be possible to guard concrete args under a different flag than record_shapes=True?

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45628410).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45628410).

**Comment by davidberard98** [comment]
@aaronenyeshi I ran some internal profiles (links provided in the internal diff stack) that show that, even in pathological cases, the slow down is almost certainly less than 2%.

The example:
```python
from time import time
import torch

@torch.jit.script
def fn(x, y):
    return (
        (x + y).as_strided([6, 2] + [1] * 27, [1, 6] + [6] * 27)
        .as_strided([1, 12] + [1] * 27, [1, 1] + [1] * 27)
        .as_strided([3, 4] + [1] * 27, [4, 1] + [4] * 27)
        .as_strided([3, 4] + [1] * 27, [1, 3] + [1] * 27)
        .as_strided([4] + [1] * 30, [1] * 31)
        .as_strided([2, 3, 2] + [1] * 26, [3, 1, 6] + [1] * 26)
    )

x, y = [torch.rand(4, 3) for _ in range(2)]

result = [[], []]

for i in range(1000):
    idx = i % 2
    concrete_inputs = (idx == 1)
    with torch.profiler.profile(record_shapes=True, record_concrete_inputs=concrete_inputs) as prof:
        begin = time()
        for _ in range(1000):
            fn(x, y)
        fin = time()

    if i >= 10:
        result[idx].append(fin-begin)
    if i % 50 == 0:
        print(f"  iteration {idx}")

print(f"No CI:{result[0]}")
print(f"With CI:{result[1]}")

prof.export_chrome_trace("/home/dberard/dberard_chrome_trace.json")
```

Over 1000 traces we get:

| Experiment  | Mean (ms) | Stdev | Stdev / sqrt(N) |
| ------------- | ------------- | ------ | ------ |
| Without concrete_inputs  | 14.73  | 0.944 | 0.0425 | 
| With concrete_inputs  | 14.68  | 0.813 | 0.0366 |

**Review by aaronenyeshi — APPROVED** [review]
This is great! Thank you!

File: torch/csrc/profiler/collection.h (hunk 1 above)
Comment (aaronenyeshi) [reply torch/csrc/profiler/collection.h:151]: We can go either way, I think the current PR looks good for now, and also your name change from inputs_ to input_shapes_ sounds looks good.

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45628410).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45628410).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45798431).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45798431).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45798431).

**Review by aaronenyeshi — APPROVED** [review]
